src/OrchardCore.Modules/OrchardCore.Search.Lucene/Services/LuceneIndexManager.cs

```diff
@@ -139,11 +140,11 @@ public bool Exists(string indexName)
 
         public async Task StoreDocumentsAsync(string indexName, IEnumerable<DocumentIndex> indexDocuments)
         {
-            await WriteAsync(indexName, writer =>
+            await WriteAsync(indexName, async writer =>
```

Comment:
Maybe but the Lucene files needs a Locking mechanism. I would need to test things to be sure. Fire and forget, I'm not sure 100%. It still needs a locking mechanism to prevent 2 threads trying to open and write the Lucene files.

Reply:
What I think @Piedone meant is that currently the passed delegate is an `Action<IndexWriter>`.

    private async Task WriteAsync(string indexName, Action<IndexWriter> action, bool close = false)

So normally you can't use the async await keywords, one solution is to remove them and in place of the await use `GetAwaiter().GetResult()` but not so good, the other solution is to declare another one like this.

    private async Task WriteAsync(string indexName, Func<IndexWriter, Task> action, bool close = false)
    {
        ...
        if (action != null)
        {
            await action(writer);
        }

Then I think better to only have one declaration, the above one, and for the other usages, e.g. this one

    await WriteAsync(indexName, _ => { }, true);

would become something like this

    await WriteAsync(indexName, _ => Task.CompletedTask, true);

Just saw that we use a `lock (this)` block in which we can't use await, so we would just need to declare and use in this tenant singleton a `SemaphoreSlim` (async lock) as we use in quite many places.

Reply:
My point is, why make it async if you need to lock it afterward with a SemaphoreSlim?

Reply:
That's okay if it is an async lock `await semaphore.WaitAsync()`, it waits to acquire a lock (if another concurrent executes it) so it is blocked in its async flow but not block the thread which can be made available to the thread pool, as any other async call, which is not the case of lock (this).

Then idem for the action if we use await actionAsync(...) which is not allowed in a regular lock block.

Reply:
The point is that if you use `async`, then it shouldn't be a fire and forget one (the sole exception being .NET-style event handlers).